### DISCUSSION TITLE
Nested Fields

### DISCUSSION THREAD

**Discussion by saultawil** [apache/superset#18435] [Ideas]
I would like to know if there is any plans to support nested fields like some other tools that now support this like elasticsearch/kibana.

I use Superset against a Spark SQL HiveServer2 compatible JDBC query engine that reads data from Hive tables that support nested data but in Superset I cannot selected the nested fields independently.

I spend a lot of time flattening fields for new projects. 
Also, for users, referencing the fields with dot notation would be much more intuitive.

A lot of modern data arrives in json with nested struct fields and is stored on an object store like S3. Since there are more and more data lakes or lakehouses front-ended with query engines like Presto and Spark with this type of nested data, users/developers want a simple way of querying this data with minimal change. So it would be very strategic and beneficial to Superset to support such a capability.

**Comment by junlincc** [comment]
@saultawil thanks for posting this enhancement request. sorry that we couldn't answer your questing in the meetup on Wednesday. I don't think we currently has the ability of running independent query on nested data, there might be work around that I don't know though. 

Tableau offers something similar to your request - after user connect to a JSON file, it prompts users to select the schema levels. Is this what you are looking for?

![renditionDownload](https://user-images.githubusercontent.com/67837651/107132691-0dbff880-6896-11eb-84bf-1f60cc32b7ad.jpeg)
![renditionDownload (1)](https://user-images.githubusercontent.com/67837651/107132694-10bae900-6896-11eb-9baa-441f40c0bf85.jpeg)

@villebro @zhaoyongjie

**Comment by saultawil** [comment]
Thanks for the response Junlin...I suppose Tableau may offer access to nested fields but the above example is more complex because it involves arrays which can be repeatable sets of nested data. I am looking for support for something much simpler - simple struct fields where one struct field is made of multiple nested fields that occur exactly once. An example that I have seen in a similar tool called Kibana (see below image attached) - I did not set up this data source but I do not think it required any special input from the user regarding levels. The tool simply recognizes the nested fields and treats them like separate fields that can be accessed with dot notation not matter the level:

![image](https://user-images.githubusercontent.com/60597083/107143410-934ab380-693d-11eb-854c-3cb5e6f45e22.png)

**Comment by DanielLyi** [comment]
Hey. Any updates regarding this issue? Are there any timeboxes on when it's going to be implemented?

**Comment by VMRuiz** [comment]
Hello, 

I'm also be interested in this feature. In my case I have an array of object with fields `key` and `value` and I need to query the value for one particular key. Currently, I'm able to access these fields through the SQL console like: 

```SQL
# Select all instances with key == foo inside the array

select id, cm.value
from `db`.`table`, unnest(custom_metrics) as cm
where cm.key = "foo";
```

It would gladly help us to have some way to perform this type of selection directly during the chart selection. Currently, doing something like this: 

![image](https://github.com/apache/superset/assets/5072090/58b47332-e1a2-4d63-8784-39e3d4011795)

causes this error:

![image](https://github.com/apache/superset/assets/5072090/70cf85c3-a72b-49bc-818f-52c315c81386)

because the field values are not being unnested first. 

Would it be possible to add a check to indicate that the given field should be unnested like in my previous query example?

**Reply by rumbin** [reply]
@VMRuiz I think your example is a bit special. These are not nested fields but an _array_ of objects. If it were a map instead of an array, this could be achieved without unnesting the data.
However, as you show in your example, your data structure requires a lateral cross join for unnesting it and this is in my eyes way too much to expect from Superset and thus out of the scope of this discussion.
An implicit lateral unnesting of arrays would require a lot of schema knowledge.

I am confident that if you convert the array of kay/value pairs to a plain map, you should be able to address the value of a given key as simple as `custom_metrics.foo` in Superset.

**Reply by VMRuiz** [reply]
@rumbin You are right, I didn't notice that.  Sorry for going out of topic then ! 

> I am confident that if you convert the array of kay/value pairs to a plain map, you should be able to address the value of a given key as simple as custom_metrics.foo in Superset.

Do you mean converting it at DB-level  or during the querying process?

**Reply by rumbin** [reply]
@VMRuiz the array-to-map conversion can be done within a virtual Dataset instead of materializing it on the DB. However, I'd only do that if the table is small, as it could be an expensive process and you'd likely not want the DB to do so on the fly for each and every refresh of a chart that uses thus dataset.
Furthermore, depending on how your write the query and how clever the query engine is, it may happen that chart or dashboard filters won't be applied _before_ doing the conversion. This would be very inefficient.
